### PR TITLE
Set travis build in stages

### PR DESCRIPTION
**Pull request by r-rojo**
in .travis.yml:
* Break out the four different build/test components into individual jobs
* Add `pip` and `ccache` cache directives to speed up provisioning of VM

in test-codingstyle-py.sh
* Use `TRAVIS_COMMIT_RANGE` env. variable to get a list of changed files
* Set the list of `FILES` to look at to the Intersection of changed files and the list of Python files

Did a few experiments with changing a Python file (`mux.py`)

### CODE REVIEW

**Comment by r-rojo** [comment]
This is what the build will look like in travis with jobs turned on:

image

File: scripts/test-codingstyle-py.sh [hunk 1]
```diff
@@ -8,11 +8,21 @@ FILES=$(find . -iname "*.py" -not -name "cpplint.py" -not -path "./doc/*")
 FILES+=" "
 FILES+=$(grep -rl "^#./usr/bin.*python" ./* | grep -v cpplint.py | grep -vE "^\.\/doc\/")
 
+if [ "$TRAVIS_COMMIT_RANGE" != "" ]; then
+    CHANGED_FILES=$(git diff --name-only $TRAVIS_COMMIT_RANGE)
+    printf "Looking at changed files:\n${CHANGED_FILES}"
+    FILES=$(comm -12 <(sort <(for f in $FILES; do printf "$f\n"; done)) <(sort <(for f in $CHANGED_FILES; do printf "./$f\n"; done)))
```
Comment (tswhison) [inline scripts/test-codingstyle-py.sh]: Say what?

Comment (r-rojo) [reply scripts/test-codingstyle-py.sh]: Use `git diff` to get a list of files changed in this push
Use `comm` to compare two files
* `<( _expression_)` will generate a temporary file
* Sort and Iterate over both lists and add a newline character

**Comment by tswhison** [comment]
Looks like it needs a rebase..

**Pull request merged**